src/app/shared/myentries.service.ts

```diff
@@ -13,121 +13,92 @@
+  constructor(protected urlResolverService: UrlResolverService) {
+    super();
+  }
 
-@Injectable()
-export abstract class MyEntriesService extends Base {
   /**
-   * Find the index for the group that the entry belongs to
-   * @param  orgWorkflows Array of grouped entries
-   * @param  group        Group to place entry in
-   * @return              The index of the entry group
+   * Gets all the user's current entries for a specific entry type (Tool, BioWorkflow, Service)
+   *
+   * @abstract
+   * @memberof MyEntriesService
    */
-  protected abstract getGroupIndex(groupEntries: any[], group: string): number;
+  abstract getMyEntries(userId: number, entryType: EntryType): void;
+  abstract registerEntry(entryType: EntryType): void;
+
+  protected findEntryFromPath(path: string | null, entries: Array<R> | null): R | null {
+    if (!path || !entries || entries.length === 0) {
+      return null;
+    }
+    return entries.find(entry => this.getPath(entry) === path);
+  }
+
+  abstract getPath(entry: DockstoreTool | Workflow): string;
 
   /**
-   * Sorts the groups of entries and returns
-   * @param  groupEntries Array of groups of entries by partial path
-   * @param  username     Dockstore username
-   * @param  type         either tool or workflow
-   * @return              A sorted array of entries grouped together
+   * Precondition: URL does not yield any useful tool
+   * Select the first published tool. If there's no published, select the first unpublished tool.
+   * @param entries
    */
-  sortGroups(groupEntries: any[], username: string, type: string): any {
-    let sortedGroupEntries = [];
-    /* User's entries Appear in First Section */
-    let unIndex = -1;
-    let orIndex = -1;
-    let orGroupObj = null;
-    for (let i = 0; i < groupEntries.length; i++) {
-      if (groupEntries[i][type] === username) {
-        unIndex = i;
-        sortedGroupEntries.push(groupEntries[i]);
-      } else if (groupEntries[i][type] === '_') {
-        orIndex = i;
-        orGroupObj = {
-          organization: 'Others',
-          namespace: 'Others',
-          entries: groupEntries[i].entries
-        };
-      }
+  protected getInitialEntry(entries: Array<R> | null): R | null {
+    if (!entries || entries.length === 0) {
+      return null;
     }
-    if (unIndex >= 0) {
-      groupEntries.splice(unIndex, 1);
+    const publishedEntries = entries.filter(entry => entry.is_published);
+    if (publishedEntries.length > 0) {
+      publishedEntries.sort(this.sortEntry);
+      return publishedEntries[0];
+    } else {
+      return entries[0];
     }
-    if (orIndex >= 0) {
-      groupEntries.splice(unIndex < orIndex ? orIndex - 1 : orIndex, 1);
+  }
+  isWorkflow(entry: DockstoreTool | Workflow): entry is Workflow {
+    return (entry as Workflow).full_workflow_path !== undefined;
+  }
+  protected sortEntry(entryA: DockstoreTool | Workflow, toolB: DockstoreTool | Workflow): number {
```

Comment:
this is super nitpicky, but entryA+ toolB? 
Should be entryA+entryB or toolA+toolB?

Reply:
oops